lom/src/pluginmgr/pluginmgr_common/pluginmgr_helper.go

```diff
@@ -115,63 +165,436 @@ func (plmgr *PluginManager) getPluginMetadata(plgname string) (plugins_common.IP
+}
+
+/*
+ * handleRequest : Handle request from server
+ * Basic validation on ActionRequestData is done. Detail validation job is responsibility of plugin
+ */
+func (plmgr *PluginManager) handleRequest(actionReq *lomipc.ActionRequestData) error {
+    // Verify request message and get plugin object for the request action
+    plugin, pluginmetadata, err := plmgr.verifyRequestMsg(actionReq)
+    if err != nil {
+        return err
+    }
+
+    lomcommon.LogInfo(
+        "In handleRequest(): Processing action request for plugin:%s, timeout:%d InstanceId:%s AnomalyInstanceId:%s AnomalyKey:%s",
+        actionReq.Action,
+        actionReq.Timeout,
+        actionReq.InstanceId,
+        actionReq.AnomalyInstanceId,
+        actionReq.AnomalyKey,
+    )
+
+    // create a channel for receiving the responses from plugin
+    respChan := make(chan *lomipc.ActionResponseData, 1)
+
+    // create a channel for receiving the heartbeats from plugin
+    hbChan := make(chan plugins_common.PluginHeartBeat, 1)
+
+    // create a channel for stopping the periodic loogging
+    var stopChan chan bool
+
+    // Start the plugin request in a separate goroutine
+    lomcommon.GetGoroutineTracker().Start("plg_mgr_handleRequest_"+actionReq.Action+"_"+lomcommon.GetUUID(),
+        func() {
+            pluginmetadata.SetPluginStage(plugins_common.PluginStageRequestStarted)
+
+            // Invoke the plugin request and send the heartbeats through the channel
+            resp := plugin.Request(hbChan, actionReq)
+
+            // Send the response back to the respChan
+            respChan <- resp
+        })
+
+    handleResponseFunc := func(respData *lomipc.ActionResponseData) {
+        if respData == nil {
+            lomcommon.LogPanic("In handleRequest(): Received nil response from plugin %v", actionReq.Action)
+        }
+
+        lomcommon.LogDebug("In handleRequest(): Received response from plugin %v", respData.Action)
+
+        if respData.Action != actionReq.Action {
+            lomcommon.LogPanic("In handleRequest(): Invalid action name received. Got  %v, expected %v",
+                respData.Action, actionReq.Action)
+        }
+
+        // For plugin with timeouts, stop the periodic log if its started due to previous request timeout
+        if actionReq.Timeout != 0 && stopChan != nil {
+            stopChan <- true
+        }
+
+        // check misbehaving plugins for long running plugins
+        if actionReq.Timeout != 0 && plmgr.handleMisbehavingPlugins(respData, pluginmetadata) {
+            return
+        }
+
+        plmgr.sendResponseToEngine(respData)
+
+        pluginmetadata.SetPluginStage(plugins_common.PluginStageRequestSuccess)
+    }
+
+    if actionReq.Timeout == 0 {
+        plmgr.handleRequestWithHeartbeats(actionReq, hbChan, respChan, handleResponseFunc) // long running
+    } else {
+        plmgr.handleRequestWithTimeouts(actionReq, respChan, handleResponseFunc) // short running
+    }
+
```

Comment:
Log info at the end so that we can check how much time plugin manager takes to handle a request ?

Reply:
Added at end